spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -100,7 +101,7 @@ case class RewriteMergeInto(spark: SparkSession) extends Rule[LogicalPlan] with
 
         AppendData.byPosition(target, writePlan, Map.empty)
 
-      case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
+      case MergeIntoTable(target@ IcebergTable(_), source: LogicalPlan, cond, matchedActions, notMatchedActions)
```

Comment:
It would be a bit shorter and easier to understand if you used `IcebergTable(target)` instead of `@`.

Can you also remove `: LogicalPlan`? Those aren't needed.